File: client-app/src/lib/routes/comments.ts
import { useAPIData, useAPIMutation } from "../api-actions";
import { applyCommentVote } from "../comment-votes";
import type { Comment, CommentThread, CommentVote } from "@/lib/types";

/** Every user's comments on a song: newest first, each with its replies oldest first. */
export function useSongComments(songId?: string) {
    const x = useAPIData<CommentThread[]>("/comments", {
        song_id: songId,
    });

    return {
        songComments: x.data,
        songCommentsLoading: x.isLoading,
        songCommentsErr: x.error,
    };
}

type NewCommentPayload = {
    song_id: string;
    content: string;
    /** the top level comment on the same song this one replies to */
    parent_id?: number;
};
/** Leaves a comment on a song, or a reply when `parent_id` is set. Resolves to the new comment. */
export function useCreateComment() {
    const x = useAPIMutation<NewCommentPayload, Comment>(
        "POST",
        "/comments",
        ({ song_id }) => [{ path: "/comments", params: { song_id } }],
    );
    return {
        createCommentErr: x.error,
        createCommentLoading: x.isMutating,
        resetCreateComment: x.reset,
        createComment: x.trigger,
    };
}

/** Deletes one of the user's comments, and every reply to it. */
export function useDeleteComment() {
    // the payload has no song id, so every song's comments revalidate
    const x = useAPIMutation<{ comment_id: number }, void>(
        "DELETE",
        "/comments",
        [{ path: "/comments" }],
    );
    return {
        deleteCommentErr: x.error,
        deleteCommentLoading: x.isMutating,
        resetDeleteComment: x.reset,
        deleteComment: x.trigger,
    };
}

type VotePayload = {
    comment_id: number;
    /** null takes the user's vote back */
    vote: CommentVote | null;
};
/**
 * Sets or takes back the user's vote on one of a song's comments. Shows the
 * vote in that song's cached comments straight away, rolls it back if the
 * request fails, and revalidates the comments either way.
 */
export function useVoteOnComment(songId?: string) {
    const comments = useAPIData<CommentThread[]>("/comments", {
        song_id: songId,
    });
    // lists nothing to invalidate, since the mutate below revalidates the comments
    const x = useAPIMutation<VotePayload, void>("POST", "/comments/votes");

    async function voteOnComment(commentId: number, vote: CommentVote | null) {
        await comments.mutate(
            async (current) => {
                await x.trigger({ comment_id: commentId, vote });
                return current && applyCommentVote(current, commentId, vote);
            },
            {
                optimisticData: (current) =>
                    applyCommentVote(current ?? [], commentId, vote),
                rollbackOnError: true,
                revalidate: true,
            },
        );
    }

    return {
        voteOnCommentErr: x.error,
        voteOnComment,
    };
}

File: backend-api/src/routes/comments.rs
use crate::{
    AppState,
    auth::SupabaseClaims,
    db::{
        comment_votes::{CommentVote, VoteTally, get_song_vote_tallies, set_comment_vote},
        comments::{delete_user_comment, get_song_comments, new_comment},
    },
    err::CadenzaError,
    routes::json::comment::{Comment, CommentThread},
};
use axum::{
    Json, Router,
    extract::{Query, State},
    routing::{delete, get, post},
};
use axum_jwt_auth::Claims;
use sea_orm::DatabaseConnection;
use serde::Deserialize;

#[derive(Deserialize)]
pub struct GetCommentsQueryParams {
    song_id: String,
}

/// Returns every comment on a song, from every user. Top level comments come
/// newest first, each with its replies oldest first. `mine` is true on the
/// comments the signed in user left, `votes` is up votes minus down votes, and
/// `my_vote` is the signed in user's vote.
///
/// JSON return value format:
/// ```json
/// [
///   {
///     "id": 41,
///     "content": "the bridge at 2:10",
///     "created_at": "2026-09-15T18:03:11.482913Z",
///     "mine": false,
///     "votes": 3,
///     "my_vote": "up",
///     "replies": [
///       { "id": 42, "content": "agreed", "created_at": "2026-09-15T18:20:05.107Z", "mine": true, "votes": 0, "my_vote": null }
///     ]
///   },
///   ...
/// ]
/// ```
async fn get_song_comments_handler(
    State(db): State<DatabaseConnection>,
    Claims { claims, .. }: Claims<SupabaseClaims>,
    Query(params): Query<GetCommentsQueryParams>,
) -> Result<Json<Vec<CommentThread>>, CadenzaError> {
    let threads = get_song_comments(&db, &params.song_id).await?;
    let tallies = get_song_vote_tallies(&db, claims.user_id, &params.song_id).await?;

    Ok(Json(
        threads
            .into_iter()
            .map(|(top_level, replies)| {
                CommentThread::new(top_level, replies, &tallies, claims.user_id)
            })
            .collect(),
    ))
}

#[derive(Deserialize)]
pub struct NewCommentPayload {
    song_id: String,
    /// the top level comment this one replies to, left out for a top level comment
    parent_id: Option<i64>,
    content: String,
}

/// Leaves the user's comment on a song, or a reply when `parent_id` is set
/// (see `db::comments::new_comment` for the checks). Returns the new comment.
///
/// JSON return value format:
/// ```json
/// { "id": 43, "content": "the bridge at 2:10", "created_at": "2026-09-15T18:03:11.482913Z", "mine": true, "votes": 0, "my_vote": null }
/// ```
async fn new_comment_handler(
    State(db): State<DatabaseConnection>,
    Claims { claims, .. }: Claims<SupabaseClaims>,
    Json(payload): Json<NewCommentPayload>,
) -> Result<Json<Comment>, CadenzaError> {
    let comment = new_comment(
        &db,
        claims.user_id,
        payload.song_id,
        payload.parent_id,
        &payload.content,
    )
    .await?;

    // a comment that was just left has no votes yet
    Ok(Json(Comment::new(
        comment,
        VoteTally::default(),
        claims.user_id,
    )))
}

#[derive(Deserialize)]
pub struct DeleteCommentPayload {
    comment_id: i64,
}

/// Deletes one of the user's comments along with every reply to it. Returns an
/// empty body, or `NotFound` if the user has no comment with that id.
async fn delete_user_comment_handler(
    State(db): State<DatabaseConnection>,
    Claims { claims, .. }: Claims<SupabaseClaims>,
    Json(payload): Json<DeleteCommentPayload>,
) -> Result<(), CadenzaError> {
    delete_user_comment(&db, claims.user_id, payload.comment_id).await
}

#[derive(Deserialize)]
pub struct VotePayload {
    comment_id: i64,
    /// "up" or "down", or null to take the user's vote back
    vote: Option<CommentVote>,
}

/// Sets the user's vote on a comment, replacing the vote they already cast, or
/// takes it back when `vote` is null. Returns an empty body, or `NotFound` if an
/// up or down vote names a comment that doesn't exist.
async fn vote_on_comment_handler(
    State(db): State<DatabaseConnection>,
    Claims { claims, .. }: Claims<SupabaseClaims>,
    Json(payload): Json<VotePayload>,
) -> Result<(), CadenzaError> {
    set_comment_vote(&db, claims.user_id, payload.comment_id, payload.vote).await
}

pub fn get_comments_router() -> Router<AppState> {
    Router::new()
        .route("/", get(get_song_comments_handler))
        .route("/", post(new_comment_handler))
        .route("/", delete(delete_user_comment_handler))
        .route("/votes", post(vote_on_comment_handler))
}
